docs/docs/reference/derivation.md

```diff
@@ -0,0 +1,391 @@
+calling the `error` method defined in `scala.compiletime`.
+
+**Note:** At the moment our error diagnostics for meta programming does not support yet interpolated string arguments for the `scala.compiletime.error` method that is called in the second case above. As an alternative, one can simply leave off the second case, then a missing typeclass would result in a "failure to reduce match" error.
+
+**Example:** Here is a slightly polished and compacted version of the code that's generated by inline expansion for the derived `Eq` instance of class `Tree`.
+
+```scala
+implicit def derived$Eq[T](implicit elemEq: Eq[T]): Eq[T] = new Eq[Tree[T]] {
+  def eql(x: Tree[T], y: Tree[T]): Boolean = {
+    val ev = implOf[Generic[Tree[T]]]
+    val mx = ev.reflect(x)
+    val my = ev.reflect(y)
+    mx.ordinal == my.ordinal && {
+      if (mx.ordinal == 0) {
+        derived$Eq.eql(mx(0).asInstanceOf[Tree[T]], my(0).asInstanceOf[Tree[T]]) &&
+        derived$Eq.eql(mx(1).asInstanceOf[Tree[T]], my(1).asInstanceOf[Tree[T]])
+      }
+      else if (mx.ordinal == 1) {
+        elemEq.eql(mx(0).asInstanceOf[T], my(0).asInstanceOf[T])
+      }
+      else throw new MatchError(mx.ordinal)
+    }
+  }
+}
+```
+
+One important difference between this approach and Scala-2 typeclass derivation frameworks such as Shapeless or Magnolia is that no automatic attempt is made to generate typeclass instances of elements recursively using the generic derivation framework. There must be an implicit instance of `Eq[T]` (which can of course be produced in turn using `Eq.derived`), or the compilation will fail. The advantage of this more restrictive approach to typeclass derivation is that it avoids uncontrolled transitive typeclass derivation by design. This keeps code sizes smaller, compile times lower, and is generally more predictable.
+
+### Derived Instances Elsewhere
+
+Sometimes one would like to derive a typeclass instance for an ADT after the ADT is defined, without being able to change the code of the ADT itself.
+To do this, simply define an instance with the `derived` method of the typeclass as right hand side. E.g, to implement `Ordering` for `Option`, define:
+```scala
+implicit def myOptionOrdering[T: Ordering]: Ordering[Option[T]] = Ordering.derived
+```
+Usually, the `Ordering.derived` clause has an implicit parameter of type
+`Generic[Option[T]]`. Since the `Option` trait has a `derives` clause,
+the necessary implicit instance is already present in the companion object of `Option`.
+If the ADT in question does not have a `derives` clause, an implicit `Generic` instance
+would still be synthesized by the compiler at the point where `derived` is called.
+This is similar to the situation with type tags or class tags: If no implicit instance
+is found, the compiler will synthesize one.
```

Comment:
Could the synthesis of `Generic` by kept as an implementation detail? Do we need to be aware of where it comes from, if anyway we can always get a `Generic[A]` when we ask one?

Reply:
Yes, we could keep it as a detail. It's just that the synthesis does generate a significant amount of code, so it's good to know that this will be done only once per class with a derives clause instead of once per usage.